### PR TITLE
Prometheus remote write endpoint

### PR DESCRIPTION
**Pull request by ChrsMark**
## What does this PR do?

This PR implements Prometheus remote write as a new metricset.

#### Start metricbeat
1. Enable metricbeat Prometheus module
2.  config:
```
- module: prometheus
  period: 10s
  metricsets: ["remote_write"]
```
3. start metricbeat and expect to see:
```
2020-02-27T17:49:14.724+0200	DEBUG	[module]	module/wrapper.go:189	Starting metricSetWrapper[module=prometheus, name=remote_write, host=]
2020-02-27T17:49:14.724+0200	INFO	remote_write/remote_write.go:92	Starting HTTP server on localhost:9201
```

#### Sending metrics from Prometheus
Use this forked project: https://github.com/ChrsMark/dockprom.
1. Download the project.
2. Add to `prometheus/prometheus.yml:
```
remote_write:
  - url: "http://host.docker.internal:9201/write"
```
3. `docker-compose up`

With that metrics should be collected by metricbeat and reported.

## Related issues

- Closes https://github.com/elastic/beats/issues/14983

## Notes about changes
This PR changes `HttpServer` so as to accept `handlerFunc` from metricset on server initialisation. This makes it reusable from different metricsets and not only `server` metricset of `http` module.  In order to confirm that  `server` metricset of `http` module is not broken I run `INTEGRATION_TESTS=1 BEAT_STRICT_PERMS=false nosetests ./module/http/test_http.py` successfully which shows that functionality is preserved.

### CODE REVIEW

**Comment by ChrsMark** [comment]
For this we will need to vendor `github.com/prometheus/prometheus/prompb`. Would this be a problem for the go-modules-porting @kvch?

**Comment by kvch** [comment]
If it is not urgent, please wait until go modules is merged. It is a huge effort to keep `go-modules` up to date, so I am pushing to get it in as soon as possible.

**Comment by ChrsMark** [comment]
> If it is not urgent, please wait until go modules is merged. It is a huge effort to keep `go-modules` up to date, so I am pushing to get it in as soon as possible.

Cool! No problem on waiting!

**Comment by ChrsMark** [comment]
@exekias any feedback at this point would be more than welcome. It is becoming quite big but we can split it in smaller PRs like so:
1. first PR to add a working metricset
2. refactoring code to utilize http helpers
3. system tests

File: metricbeat/module/prometheus/prometheus_pb2.py [hunk 1]
```diff
@@ -0,0 +1,577 @@
+# -*- coding: utf-8 -*-
```
Comment (exekias) [inline metricbeat/module/prometheus/prometheus_pb2.py]: is this file needed?

File: metricbeat/module/prometheus/_meta/config.yml [hunk 2]
```diff
@@ -12,3 +13,12 @@
   #bearer_token_file: /var/run/secrets/kubernetes.io/serviceaccount/token
   #ssl.certificate_authorities:
   #  - /var/run/secrets/kubernetes.io/serviceaccount/service-ca.crt
+
+- module: prometheus
```
Comment (exekias) [inline metricbeat/module/prometheus/_meta/config.yml]: will need a comment at some point

File: metricbeat/module/prometheus/remote_write/data.go [hunk 3]
```diff
@@ -0,0 +1,98 @@
...
+	labels    common.MapStr
+	timestamp time.Time
+}
+
+// LabelsHash returns a repeatable string that is unique for the set of labels in this event
+func (p *PromEvent) LabelsHash() string {
+	return p.labels.String()
+}
+
+func samplesToEvents(metrics model.Samples) map[string]mb.Event {
```
Comment (exekias) [inline metricbeat/module/prometheus/remote_write/data.go]: I understand Samples will always be untyped right? For example, we won't be able to differenciate a histogram metric from a gauge

File: metricbeat/module/prometheus/remote_write/data.go [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+			if len(promEvent.labels) > 0 {
+				eventList[labelsHash].ModuleFields["labels"] = promEvent.labels
+			}
+		}
+
+		// Not checking anything here because we create these maps some lines before
+		e := eventList[labelsHash]
+		e.Timestamp = promEvent.timestamp
+		e.ModuleFields["metrics"].(common.MapStr).Update(promEvent.data)
+	}
```
Comment (exekias) [inline metricbeat/module/prometheus/remote_write/data.go]: this can probably be optimized, by storing events already indexed by label hash, so only an insert is needed

File: metricbeat/helper/server/http/http.go [hunk 5]
```diff
@@ -44,6 +44,12 @@ type HttpEvent struct {
 	meta  server.Meta
 }
 
+// MetricSetFactory accepts a BaseMetricSet and returns a MetricSet. If there
+// was an error creating the MetricSet then an error will be returned. The
+// returned MetricSet must also implement either EventFetcher or EventsFetcher
+// (but not both).
+type HandlerFactory func(h *HttpServer) func(writer http.ResponseWriter, req *http.Request)
```
Comment (exekias) [inline metricbeat/helper/server/http/http.go]: Do you really need a factory? The metricset is already initializing the server, so you could pass a private `metricset.handleFunc` which implements `func(writer http.ResponseWriter, req *http.Request)`

File: metricbeat/helper/server/http/http.go [hunk 6]
```diff
@@ -52,7 +58,15 @@ func (h *HttpEvent) GetMeta() server.Meta {
 	return h.meta
 }
 
-func NewHttpServer(mb mb.BaseMetricSet) (server.Server, error) {
+func (h *HttpEvent) SetEvent(event common.MapStr) {
+	h.event = event
+}
```
Comment (exekias) [inline metricbeat/helper/server/http/http.go]: Having that you are now passing a custom handler when initializing (which I think is a good idea), do you need an HttpEvent construct anymore? This was used to pass data between a generic HTTP server and the metricset. But now the metricset is in control of the requests, and can directly build and send events to the output. That should make things simpler, WDYT?

File: metricbeat/module/prometheus/prometheus_pb2.py (hunk 1 above)
Comment (ChrsMark) [reply metricbeat/module/prometheus/prometheus_pb2.py]: It is required by the python tests: https://github.com/elastic/beats/pull/16609/files#diff-f0fbea18a4681afd920cb91b903d03d2R6

File: metricbeat/module/prometheus/remote_write/data.go (hunk 3 above)
Comment (ChrsMark) [reply metricbeat/module/prometheus/remote_write/data.go]: Prometheus sends data with [writerequest](https://godoc.org/github.com/prometheus/prometheus/prompb#WriteRequest) using this [proto](https://github.com/prometheus/prometheus/blob/master/prompb/remote.proto). Not sure if we can retrieve the type of the metric in an easy way 🤔

**Comment by ChrsMark** [comment]
In order to confirm that  `server` metricset of `http` module is not broken I run `INTEGRATION_TESTS=1 BEAT_STRICT_PERMS=false nosetests ./module/http/test_http.py` successfully which shows that functionality is preserved.

File: metricbeat/module/prometheus/docker-compose.yml [hunk 7]
```diff
@@ -9,3 +9,12 @@ services:
...
       - 9090
+  prometheus-host-network:
+    image: docker.elastic.co/integrations-ci/beats-prometheus:${PROMETHEUS_VERSION:-2.6.0}-1
+    build:
+      context: ./_meta
+      args:
+        PROMETHEUS_VERSION: ${PROMETHEUS_VERSION:-2.6.0}
+    ports:
+      - 9090
+    network_mode: host
```
Comment (jsoriano) [inline metricbeat/module/prometheus/docker-compose.yml]: Please extend `prometheus` service to reduce duplication.
```suggestion
  prometheus-host-network:
    extends:
      service: prometheus
    network_mode: host
```

File: metricbeat/module/prometheus/docker-compose.yml [hunk 8]
```diff
@@ -9,3 +9,12 @@ services:
         PROMETHEUS_VERSION: ${PROMETHEUS_VERSION:-2.6.0}
     ports:
       - 9090
+  prometheus-host-network:
```
Comment (jsoriano) [inline metricbeat/module/prometheus/docker-compose.yml]: Add a comment about when this service is needed in the host network.

File: metricbeat/helper/server/http/http_test.go [hunk 9]
```diff
@@ -32,6 +32,9 @@ import (
 	"testing"
 	"time"
 
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/logp"
```
Comment (jsoriano) [inline metricbeat/helper/server/http/http_test.go]: Nit. Imports order.

File: metricbeat/helper/server/http/http.go [hunk 10]
```diff
@@ -19,40 +19,24 @@ package http
...
-func (h *HttpEvent) GetEvent() common.MapStr {
-	return h.event
-}
-
-func (h *HttpEvent) GetMeta() server.Meta {
-	return h.meta
-}
-
-func NewHttpServer(mb mb.BaseMetricSet) (server.Server, error) {
+func NewHttpServer(mb mb.BaseMetricSet, handlerFunc func(writer http.ResponseWriter, req *http.Request)) (server.Server, error) {
```
Comment (jsoriano) [inline metricbeat/helper/server/http/http.go]: Use `http.HandlerFunc` already here?
```suggestion
func NewHttpServer(mb mb.BaseMetricSet, handlerFunc http.HandlerFunc) (server.Server, error) {
```

File: metricbeat/helper/server/http/http.go [hunk 11]
```diff
@@ -109,50 +91,8 @@ func (h *HttpServer) Stop() {
...
-	case "GET":
-		writer.WriteHeader(http.StatusOK)
-		if req.TLS != nil {
-			writer.Write([]byte("HTTPS Server accepts data via POST"))
-		} else {
-			writer.Write([]byte("HTTP Server accepts data via POST"))
-		}
-
-	}
+	return make(chan server.Event)
```
Comment (jsoriano) [inline metricbeat/helper/server/http/http.go]: I guess this method can be removed now?

File: metricbeat/helper/server/http/http.go (hunk 10 above)
Comment (jsoriano) [inline metricbeat/helper/server/http/http.go]: This change would need a changelog entry in the developers changelog. Or we could keep this method and add a new `NewHttpServerWithHandler`.

File: metricbeat/helper/server/http/http_test.go [hunk 12]
```diff
@@ -88,6 +95,45 @@ func TestHTTPServers(t *testing.T) {
 	}
 }
 
+func getHandleFunc(events chan common.MapStr) func(writer http.ResponseWriter, req *http.Request) {
```
Comment (jsoriano) [inline metricbeat/helper/server/http/http_test.go]: As this is a handler function for testing maybe we can simplify it, and we could use `t.Log` for logging instead of `logp`.

File: metricbeat/helper/server/http/http_test.go [hunk 13]
```diff
@@ -120,17 +166,16 @@ func checkServerReady(host string, port int) error {
 
 }
 
-func getHTTPServer(t *testing.T, host string, port int, connectionType string) (server.Server, error) {
+func getHTTPServer(t *testing.T, host string, port int, connectionType string, handleFunc func(writer http.ResponseWriter, req *http.Request)) (server.Server, error) {
```
Comment (jsoriano) [inline metricbeat/helper/server/http/http_test.go]:
```suggestion
func getHTTPServer(t *testing.T, host string, port int, connectionType string, handleFunc http.HandleFunc) (server.Server, error) {
```

File: metricbeat/module/http/server/data.go [hunk 14]
```diff
@@ -59,21 +59,21 @@ func (m *metricProcessor) RemovePath(path PathConfig) {
 	m.Unlock()
 }
 
-func (p *metricProcessor) Process(event server.Event) (common.MapStr, error) {
```
Comment (jsoriano) [inline metricbeat/module/http/server/data.go]: Why is this change needed?

File: metricbeat/module/http/server/server.go [hunk 15]
```diff
@@ -19,9 +19,15 @@ package server
 
 import (
 	"fmt"
+	"io/ioutil"
+	"net/http"
+
+	"github.com/elastic/beats/v7/libbeat/common"
+	"github.com/elastic/beats/v7/libbeat/logp"
+	"github.com/elastic/beats/v7/metricbeat/helper/server"
```
Comment (jsoriano) [inline metricbeat/module/http/server/server.go]:
```suggestion
```

File: metricbeat/module/prometheus/_meta/prometheus.yml [hunk 16]
```diff
@@ -0,0 +1,32 @@
...
+    - targets:
+      # - alertmanager:9093
+
+# Load rules once and periodically evaluate them according to the global 'evaluation_interval'.
+rule_files:
+  # - "first_rules.yml"
+  # - "second_rules.yml"
+
+remote_write:
+  - url: "http://REMOTE/write"
```
Comment (jsoriano) [inline metricbeat/module/prometheus/_meta/prometheus.yml]: Is it a problem if nothing is listening on this remote?

File: metricbeat/module/prometheus/test_prometheus.py [hunk 17]
```diff
@@ -35,3 +36,34 @@ def test_stats(self):
...
+    def test_remote_write(self):
+        """
+        prometheus remote_write test
+        """
+        self.render_config_template(modules=[{
+            "name": "prometheus",
+            "metricsets": ["remote_write"],
+            "period": "5s",
+            "host": "localhost",
+            "port": "9201"
```
Comment (jsoriano) [inline metricbeat/module/prometheus/test_prometheus.py]: I wonder if we should use the `hosts` option to configure the listener as the `http.server` metricset does. Though the `statsd` module does it this way.

File: metricbeat/module/prometheus/remote_write/remote_write.go [hunk 18]
```diff
@@ -0,0 +1,127 @@
...
+	"github.com/prometheus/common/model"
+	"github.com/prometheus/prometheus/prompb"
+
+	serverhelper "github.com/elastic/beats/v7/metricbeat/helper/server"
+	httpserver "github.com/elastic/beats/v7/metricbeat/helper/server/http"
+	"github.com/elastic/beats/v7/metricbeat/mb"
+)
+
+func init() {
+	mb.Registry.MustAddMetricSet("prometheus", "remote_write", New)
```
Comment (jsoriano) [inline metricbeat/module/prometheus/remote_write/remote_write.go]: If we don't use the `hosts` option to configure the listener we should explicitly disable it to avoid confusions.
```suggestion
	mb.Registry.MustAddMetricSet("prometheus", "remote_write", New,
	  mb.WithHostParser(parse.EmptyHostParser),
	)
```

File: metricbeat/module/prometheus/_meta/prometheus.yml (hunk 16 above)
Comment (ChrsMark) [reply metricbeat/module/prometheus/_meta/prometheus.yml]: I saw that Prometheus will just keep reporting that was not able to send data the `remote_write`.

File: metricbeat/helper/server/http/http.go (hunk 10 above)
Comment (ChrsMark) [reply metricbeat/helper/server/http/http.go]: I like the idea of `NewHttpServerWithHandler`. I will adopt it, thanks!

File: metricbeat/helper/server/http/http.go (hunk 11 above)
Comment (ChrsMark) [reply metricbeat/helper/server/http/http.go]: will not be an issue with `NewHttpServerWithHandler`

File: metricbeat/helper/server/http/http_test.go (hunk 12 above)
Comment (ChrsMark) [reply metricbeat/helper/server/http/http_test.go]: will not be an issue with `NewHttpServerWithHandler`

File: metricbeat/helper/server/http/http_test.go (hunk 13 above)
Comment (ChrsMark) [reply metricbeat/helper/server/http/http_test.go]: will not be an issue with `NewHttpServerWithHandler`

File: metricbeat/module/http/server/data.go (hunk 14 above)
Comment (ChrsMark) [reply metricbeat/module/http/server/data.go]: will not be an issue with `NewHttpServerWithHandler`

File: metricbeat/module/http/server/server.go (hunk 15 above)
Comment (ChrsMark) [reply metricbeat/module/http/server/server.go]: will not be an issue with `NewHttpServerWithHandler`

File: metricbeat/module/prometheus/test_prometheus.py (hunk 17 above)
Comment (ChrsMark) [reply metricbeat/module/prometheus/test_prometheus.py]: I think it is hardcoded there too -> https://github.com/elastic/beats/blob/master/metricbeat/module/http/test_http.py#L53
🤔

Comment (jsoriano) [reply metricbeat/module/prometheus/test_prometheus.py]: Oh, you are right, I was looking at the test for the json metricset. :+1:

File: metricbeat/helper/server/http/http.go [hunk 19]
```diff
@@ -84,6 +84,33 @@ func NewHttpServer(mb mb.BaseMetricSet) (server.Server, error) {
 	return h, nil
 }
```
Comment (jsoriano) [inline metricbeat/helper/server/http/http.go]: Could `NewHttpServer` reuse `NewHttpServerWithHandler` so we don't repeat code?

File: metricbeat/helper/server/http/http.go [hunk 20]
```diff
@@ -84,6 +84,33 @@ func NewHttpServer(mb mb.BaseMetricSet) (server.Server, error) {
...
+	config := defaultHttpConfig()
+	err := mb.Module().UnpackConfig(&config)
+	if err != nil {
+		return nil, err
+	}
+	tlsConfig, err := tlscommon.LoadTLSServerConfig(config.TLS)
+	if err != nil {
+		return nil, err
+	}
+	ctx, cancel := context.WithCancel(context.Background())
```
Comment (jsoriano) [inline metricbeat/helper/server/http/http.go]: Nit. These `context.Background()` should be `context.TODO()`.

File: metricbeat/module/prometheus/remote_write/data.go [hunk 21]
```diff
@@ -0,0 +1,74 @@
...
+	for _, metric := range metrics {
+		labels := common.MapStr{}
+
+		if metric == nil {
+			continue
+		}
+		name := string(metric.Metric["__name__"])
+		delete(metric.Metric, "__name__")
+
+		metric.Metric.Fingerprint()
```
Comment (jsoriano) [inline metricbeat/module/prometheus/remote_write/data.go]: We are not getting the output of `Fingerprint()` :thinking:

File: metricbeat/module/prometheus/remote_write/data.go (hunk 3 above)
Comment (jsoriano) [reply metricbeat/module/prometheus/remote_write/data.go]: I guess this is going to be an issue with histograms? Should we do a best-effort guess if it includes metrics with some special labels and/or suffixes?

**Comment by ChrsMark** [comment]
Comments addressed. Let's see if CI is happy.

File: metricbeat/module/prometheus/remote_write/data.go (hunk 21 above)
Comment (ChrsMark) [reply metricbeat/module/prometheus/remote_write/data.go]: 👍 this a leftover. Shouldn't be here.

**Comment by ChrsMark** [comment]
@jsoriano regarding https://github.com/elastic/beats/pull/16609#discussion_r390489240, this is something we discussed with @exekias already and guessing would be the way to go when we will decide to adopt histograms for this metricset.

I think that comments have been addressed 🙂 .

File: metricbeat/module/prometheus/_meta/run.sh [hunk 22]
```diff
@@ -1,12 +1,20 @@
...
+for i in 1 2 3 4 5;
+do
+  a=`nslookup host.docker.internal | grep "** server can't find " | wc -l`;
+  if [ $a -gt 0 ]; then
+    # this works only on Linux envs
+    HOST_DOMAIN="0.0.0.0"
+  else
+    # this works only on Mac envs
+    HOST_DOMAIN="host.docker.internal"
+    break
```
Comment (jsoriano) [inline metricbeat/module/prometheus/_meta/run.sh]: Why only breaking in this case?

Comment (ChrsMark) [reply metricbeat/module/prometheus/_meta/run.sh]: We want to keep trying to reach `host.docker.internal` otherwise to fallback to `0.0.0.0`.
